File: sdks/typescript/src/core/awareness-threshold.ts
/**
 * DCP-09 v2.0 Awareness Threshold Engine.
 *
 * Configures and evaluates when an agent must notify its human principal.
 * Mirrors the security-tier pattern but operates on significance rather than risk.
 *
 * Significance scoring: 0–1000 (millipoints).
 */

import type { CompositeKeyPair } from './composite-ops.js';
import { compositeSign } from './composite-ops.js';
import type { AlgorithmRegistry } from './crypto-registry.js';
import { canonicalizeV2 } from './canonicalize.js';
import { DCP_CONTEXTS } from './domain-separation.js';
import type {
  ThresholdOperator,
  ThresholdAction,
  ThresholdRule,
  AwarenessThreshold,
  AdvisoryDeclaration,
} from '../types/v2.js';

// ── Significance Scoring ──

export interface SignificanceContext {
  financial_impact?: number;
  data_sensitivity?: number;
  relationship_impact?: number;
  irreversibility?: number;
  precedent_setting?: number;
}

const SIGNIFICANCE_WEIGHTS = {
  financial_impact: 0.25,
  data_sensitivity: 0.20,
  relationship_impact: 0.20,
  irreversibility: 0.20,
  precedent_setting: 0.15,
} as const;

/**
 * Evaluate the significance of an action (0–1000).
 */
export function evaluateSignificance(context: SignificanceContext): number {
  let total = 0;
  for (const [key, weight] of Object.entries(SIGNIFICANCE_WEIGHTS)) {
    const value = context[key as keyof SignificanceContext] ?? 0;
    total += Math.max(0, Math.min(1, value)) * weight;
  }
  return Math.round(total * 1000);
}

// ── Threshold Evaluation ──

function evaluateOperator(operator: ThresholdOperator, actual: number, threshold: number): boolean {
  switch (operator) {
    case 'gt': return actual > threshold;
    case 'lt': return actual < threshold;
    case 'gte': return actual >= threshold;
    case 'lte': return actual <= threshold;
    case 'eq': return actual === threshold;
  }
}

/**
 * Determine whether a human should be notified given significance and threshold rules.
 */
export function shouldNotifyHuman(
  significance: number,
  thresholds: ThresholdRule[],
): { notify: boolean; triggered_rules: ThresholdRule[]; actions: ThresholdAction[] } {
  const triggered: ThresholdRule[] = [];
  const actions = new Set<ThresholdAction>();

  for (const rule of thresholds) {
    const value = rule.dimension === 'significance' ? significance : 0;
    if (evaluateOperator(rule.operator, value, rule.value)) {
      triggered.push(rule);
      actions.add(rule.action_if_triggered);
    }
  }

  return {
    notify: triggered.length > 0,
    triggered_rules: triggered,
    actions: [...actions],
  };
}

// ── Artifact Creation ──

/**
 * Create an awareness threshold configuration.
 */
export async function createAwarenessThreshold(
  registry: AlgorithmRegistry,
  keys: CompositeKeyPair,
  params: {
    threshold_id: string;
    session_nonce: string;
    agent_id: string;
    human_id: string;
    threshold_rules: ThresholdRule[];
  },
): Promise<AwarenessThreshold> {
  const payload = {
    dcp_version: '2.0' as const,
    threshold_id: params.threshold_id,
    session_nonce: params.session_nonce,
    agent_id: params.agent_id,
    human_id: params.human_id,
    threshold_rules: params.threshold_rules,
    timestamp: new Date().toISOString(),
  };

  const canonical = canonicalizeV2(payload);
  const payloadBytes = new TextEncoder().encode(canonical);
  const compositeSig = await compositeSign(registry, DCP_CONTEXTS.Awareness, payloadBytes, keys);

  return { ...payload, composite_sig: compositeSig };
}

/**
 * Create an advisory declaration to notify the human.
 */
export async function createAdvisoryDeclaration(
  registry: AlgorithmRegistry,
  keys: CompositeKeyPair,
  params: {
    declaration_id: string;
    session_nonce: string;
    agent_id: string;
    human_id: string;
    significance_score: number;
    action_summary: string;
    recommended_response: string;
    response_deadline: string;
  },
): Promise<AdvisoryDeclaration> {
  const payload = {
    dcp_version: '2.0' as const,
    declaration_id: params.declaration_id,
    session_nonce: params.session_nonce,
    agent_id: params.agent_id,
    human_id: params.human_id,
    significance_score: params.significance_score,
    action_summary: params.action_summary,
    recommended_response: params.recommended_response,
    response_deadline: params.response_deadline,
    human_response: null,
    proceeded_without_response: false,
    timestamp: new Date().toISOString(),
  };

  const canonical = canonicalizeV2(payload);
  const payloadBytes = new TextEncoder().encode(canonical);
  const compositeSig = await compositeSign(registry, DCP_CONTEXTS.Awareness, payloadBytes, keys);

  return { ...payload, composite_sig: compositeSig };
}

File: sdks/python/dcp_ai/v2/awareness_threshold.py
"""
DCP-09 v2.0 Awareness Threshold Engine — Python port.

Configures and evaluates when an agent must notify its human principal.
Significance is scored in 0..=1000 (millipoints).
"""

from __future__ import annotations

from typing import Any

from dcp_ai.v2.canonicalize import canonicalize_v2
from dcp_ai.v2.composite_ops import CompositeKeyInfo, composite_sign
from dcp_ai.v2.crypto_registry import AlgorithmRegistry
from dcp_ai.v2.domain_separation import DCP_CONTEXTS
from dcp_ai.v2.lifecycle import _utc_now

_SIGNIFICANCE_WEIGHTS: dict[str, float] = {
    "financial_impact": 0.25,
    "data_sensitivity": 0.20,
    "relationship_impact": 0.20,
    "irreversibility": 0.20,
    "precedent_setting": 0.15,
}


def evaluate_significance(context: dict[str, float]) -> int:
    """Evaluate the significance of an action in 0..=1000."""
    total = 0.0
    for key, weight in _SIGNIFICANCE_WEIGHTS.items():
        v = float(context.get(key, 0.0))
        v = max(0.0, min(1.0, v))
        total += v * weight
    return round(total * 1000)


def _evaluate_operator(operator: str, actual: float, threshold: float) -> bool:
    if operator == "gt":
        return actual > threshold
    if operator == "lt":
        return actual < threshold
    if operator == "gte":
        return actual >= threshold
    if operator == "lte":
        return actual <= threshold
    if operator == "eq":
        return actual == threshold
    raise ValueError(f"Unknown threshold operator: {operator}")


def should_notify_human(
    significance: float,
    thresholds: list[dict[str, Any]],
) -> dict[str, Any]:
    """Determine whether a human should be notified given significance + rules.

    Each rule has: dimension, operator, value, action_if_triggered.
    Returns: {notify, triggered_rules, actions}.
    """
    triggered: list[dict[str, Any]] = []
    actions: list[str] = []
    for rule in thresholds:
        value = significance if rule.get("dimension") == "significance" else 0
        if _evaluate_operator(rule["operator"], value, rule["value"]):
            triggered.append(rule)
            action = rule.get("action_if_triggered")
            if action and action not in actions:
                actions.append(action)
    return {
        "notify": bool(triggered),
        "triggered_rules": triggered,
        "actions": actions,
    }


async def create_awareness_threshold(
    registry: AlgorithmRegistry,
    classical_key: CompositeKeyInfo,
    pq_key: CompositeKeyInfo,
    *,
    threshold_id: str,
    session_nonce: str,
    agent_id: str,
    human_id: str,
    threshold_rules: list[dict[str, Any]],
) -> dict[str, Any]:
    """Create an awareness threshold configuration."""
    payload: dict[str, Any] = {
        "dcp_version": "2.0",
        "threshold_id": threshold_id,
        "session_nonce": session_nonce,
        "agent_id": agent_id,
        "human_id": human_id,
        "threshold_rules": threshold_rules,
        "timestamp": _utc_now(),
    }
    canonical = canonicalize_v2(payload)
    composite = await composite_sign(
        registry, DCP_CONTEXTS["Awareness"], canonical.encode("utf-8"), classical_key, pq_key
    )
    return {**payload, "composite_sig": composite}


async def create_advisory_declaration(
    registry: AlgorithmRegistry,
    classical_key: CompositeKeyInfo,
    pq_key: CompositeKeyInfo,
    *,
    declaration_id: str,
    session_nonce: str,
    agent_id: str,
    human_id: str,
    significance_score: int,
    action_summary: str,
    recommended_response: str,
    response_deadline: str,
) -> dict[str, Any]:
    """Create an advisory declaration to notify the human."""
    payload: dict[str, Any] = {
        "dcp_version": "2.0",
        "declaration_id": declaration_id,
        "session_nonce": session_nonce,
        "agent_id": agent_id,
        "human_id": human_id,
        "significance_score": significance_score,
        "action_summary": action_summary,
        "recommended_response": recommended_response,
        "response_deadline": response_deadline,
        "human_response": None,
        "proceeded_without_response": False,
        "timestamp": _utc_now(),
    }
    canonical = canonicalize_v2(payload)
    composite = await composite_sign(
        registry, DCP_CONTEXTS["Awareness"], canonical.encode("utf-8"), classical_key, pq_key
    )
    return {**payload, "composite_sig": composite}
